File: src/main/java/gg/mineral/practice/inventory/menus/CreateCustomKnockbackMenu.java
package gg.mineral.practice.inventory.menus;

import gg.mineral.practice.inventory.ClickCancelled;
import gg.mineral.practice.inventory.Menu;
import gg.mineral.practice.inventory.PracticeMenu;
import gg.mineral.practice.match.CustomKnockback;
import gg.mineral.practice.util.items.ItemStacks;
import gg.mineral.practice.util.messages.CC;
import gg.mineral.server.combat.KnockbackProfile;
import gg.mineral.server.combat.KnockbackProfileList;
import lombok.RequiredArgsConstructor;
import lombok.val;

import java.math.RoundingMode;
import java.text.DecimalFormat;

@ClickCancelled(true)
@RequiredArgsConstructor
public class CreateCustomKnockbackMenu extends PracticeMenu {
    private final static DecimalFormat DECIMAL_FORMAT = new DecimalFormat("#.###");

    static {
        DECIMAL_FORMAT.setRoundingMode(RoundingMode.HALF_DOWN);
    }

    private final Menu menu, prevMenu;
    private final KnockbackProfile[] profiles = KnockbackProfileList.getProfiles().values()
            .stream()
            .sorted((profile1, profile2) -> {
                if ("default_kb".equals(profile1.getName())) return -1;
                if ("default_kb".equals(profile2.getName())) return 1;
                return 0;
            })
            .toArray(KnockbackProfile[]::new);
    private int kbIndex = 0;
    private boolean oldCombat;
    private CustomKnockback customKnockback;

    public void setFriction(double friction) {
        kbIndex = -1;
        customKnockback.setFriction(friction);
    }

    public void setHorizontal(double horizontal) {
        kbIndex = -1;
        customKnockback.setHorizontal(horizontal);
    }

    public void setVertical(double vertical) {
        kbIndex = -1;
        customKnockback.setVertical(vertical);
    }

    public void setHorizontalExtra(double horizontalExtra) {
        kbIndex = -1;
        customKnockback.setHorizontalExtra(horizontalExtra);
    }

    public void setVerticalExtra(double verticalExtra) {
        kbIndex = -1;
        customKnockback.setVerticalExtra(verticalExtra);
    }

    public void setVerticalLimit(double verticalLimit) {
        kbIndex = -1;
        customKnockback.setVerticalLimit(verticalLimit);
    }

    public double getFriction(KnockbackProfile kb) {
        if (kbIndex > 0)
            return (double) kb.getConfigValues().getOrDefault("friction", customKnockback.getFriction());
        return customKnockback.getFriction();
    }

    public double getHorizontal(KnockbackProfile kb) {
        if (kbIndex > 0)
            return (double) kb.getConfigValues().getOrDefault("horizontal", customKnockback.getHorizontal());
        return customKnockback.getHorizontal();
    }

    public double getVertical(KnockbackProfile kb) {
        if (kbIndex > 0)
            return (double) kb.getConfigValues().getOrDefault("vertical", customKnockback.getVertical());
        return customKnockback.getVertical();
    }

    public double getHorizontalExtra(KnockbackProfile kb) {
        if (kbIndex > 0)
            return (double) kb.getConfigValues().getOrDefault("horizontalExtra", customKnockback.getHorizontalExtra());
        return customKnockback.getHorizontalExtra();
    }

    public double getVerticalExtra(KnockbackProfile kb) {
        if (kbIndex > 0)
            return (double) kb.getConfigValues().getOrDefault("verticalExtra", customKnockback.getVerticalExtra());
        return customKnockback.getVerticalExtra();
    }

    public double getVerticalLimit(KnockbackProfile kb) {
        if (kbIndex > 0)
            return (double) kb.getConfigValues().getOrDefault("verticalLimit", customKnockback.getVerticalLimit());
        return customKnockback.getVerticalLimit();
    }

    @Override
    public void update() {

        val duelSettings = viewer.getDuelSettings();

        for (int i = 0; i <= profiles.length; i++) {
            if (kbIndex >= 0 && profiles[kbIndex].getName().contains("combo_kb")) {
                kbIndex++;
                if (kbIndex >= profiles.length || kbIndex < 0)
                    kbIndex = 0;
                continue;
            }
            break;
        }

        val kb = kbIndex >= 0 && kbIndex < profiles.length ? profiles[kbIndex] : null;

        if (customKnockback == null) {
            customKnockback = duelSettings.getKnockback() != null
                    && duelSettings.getKnockback() instanceof CustomKnockback customKB
                    ? new CustomKnockback(customKB)
                    : new CustomKnockback();

            if (kbIndex >= 0 && kbIndex < profiles.length) {
                val friction = profiles[kbIndex].getConfigValues().getOrDefault("friction", customKnockback.getFriction());
                val horizontal = profiles[kbIndex].getConfigValues().getOrDefault("horizontal", customKnockback.getHorizontal());
                val vertical = profiles[kbIndex].getConfigValues().getOrDefault("vertical", customKnockback.getVertical());
                val horizontalExtra = profiles[kbIndex].getConfigValues().getOrDefault("horizontalExtra", customKnockback.getHorizontalExtra());
                val verticalExtra = profiles[kbIndex].getConfigValues().getOrDefault("verticalExtra", customKnockback.getVerticalExtra());
                val verticalLimit = profiles[kbIndex].getConfigValues().getOrDefault("verticalLimit", customKnockback.getVerticalLimit());
                val oldCombat = profiles[kbIndex].getConfigValues().getOrDefault("oldCombat", duelSettings.isOldCombat());
                customKnockback.setFriction((double) friction);
                customKnockback.setHorizontal((double) horizontal);
                customKnockback.setVertical((double) vertical);
                customKnockback.setHorizontalExtra((double) horizontalExtra);
                customKnockback.setVerticalExtra((double) verticalExtra);
                customKnockback.setVerticalLimit((double) verticalLimit);
                this.oldCombat = (boolean) oldCombat;
            }
        }

        setSlot(10, ItemStacks.FRICTION.name(CC.SECONDARY + CC.B + "Friction")
                        .lore(CC.WHITE + "The amount " + CC.SECONDARY + "movement speed and direction",
                                CC.WHITE + "influences knockback magnitude.", " ",
                                CC.WHITE + "Currently:",
                                CC.GOLD + DECIMAL_FORMAT.format(getFriction(customKnockback)),
                                CC.BOARD_SEPARATOR, CC.ACCENT + "Click to change value.")
                        .build(),
                interaction -> interaction.getProfile()
                        .openMenu(ConfigureValueMenu.of(this,
                                this::setFriction, double.class)));

        setSlot(11, ItemStacks.HORIZONTAL.name(CC.SECONDARY + CC.B + "Horizontal")
                        .lore(CC.WHITE + "The " + CC.SECONDARY + "horizontal" + CC.WHITE + " knockback.", " ",
                                CC.WHITE + "Currently:",
                                CC.GOLD + DECIMAL_FORMAT.format(getHorizontal(customKnockback)),
                                CC.BOARD_SEPARATOR, CC.ACCENT + "Click to change value.")
                        .build(),
                interaction -> interaction.getProfile()
                        .openMenu(ConfigureValueMenu.of(this,
                                this::setHorizontal,
                                double.class)));

        setSlot(12, ItemStacks.VERTICAL.name(CC.SECONDARY + CC.B + "Vertical")
                        .lore(CC.WHITE + "The " + CC.SECONDARY + "vertical" + CC.WHITE + " knockback.", " ",
                                CC.WHITE + "Currently:",
                                CC.GOLD + DECIMAL_FORMAT.format(getVertical(customKnockback)),
                                CC.BOARD_SEPARATOR, CC.ACCENT + "Click to change value.")
                        .build(),
                interaction -> interaction.getProfile()
                        .openMenu(ConfigureValueMenu.of(this,
                                this::setVertical, double.class)));

        setSlot(13, ItemStacks.EXTRA_HORIZONTAL
                        .name(CC.SECONDARY + CC.B + "Extra Horizontal")
                        .lore(CC.WHITE + "The " + CC.SECONDARY + "horizontal knockback" + CC.WHITE + " added",
                                CC.WHITE + "when sprinting/sprint resetting.", " ",
                                CC.WHITE + "Currently:",
                                CC.GOLD + DECIMAL_FORMAT.format(getHorizontalExtra(customKnockback)),
                                CC.BOARD_SEPARATOR, CC.ACCENT + "Click to change value.")
                        .build(),
                interaction -> interaction.getProfile()
                        .openMenu(ConfigureValueMenu.of(this,
                                this::setHorizontalExtra,
                                double.class)));

        setSlot(14, ItemStacks.EXTRA_VERTICAL.name(CC.SECONDARY + CC.B + "Extra Vertical")
                        .lore(CC.WHITE + "The " + CC.SECONDARY + "vertical knockback" + CC.WHITE + " added",
                                CC.WHITE + "when sprinting/sprint resetting.", " ",
                                CC.WHITE + "Currently:",
                                CC.GOLD + DECIMAL_FORMAT.format(getVerticalExtra(customKnockback)),
                                CC.BOARD_SEPARATOR, CC.ACCENT + "Click to change value.")
                        .build(),
                interaction -> interaction.getProfile()
                        .openMenu(ConfigureValueMenu.of(this,
                                this::setVerticalExtra,
                                double.class)));

        setSlot(15, ItemStacks.VERTICAL_LIMIT.name(CC.SECONDARY + CC.B + "Vertical Limit")
                        .lore(CC.WHITE + "The " + CC.SECONDARY + "limit" + CC.WHITE + " to vertical knockback.",
                                " ",
                                CC.WHITE + "Currently:",
                                CC.GOLD + DECIMAL_FORMAT.format(getVerticalLimit(customKnockback)),
                                CC.BOARD_SEPARATOR, CC.ACCENT + "Click to change value.")
                        .build(),
                interaction -> interaction.getProfile()
                        .openMenu(ConfigureValueMenu.of(this,
                                this::setVerticalLimit,
                                double.class)));

        setSlot(16, ItemStacks.OLD_COMBAT.name(CC.SECONDARY + CC.B + "Delayed Combat").lore(
                        CC.WHITE + "Play using " + CC.SECONDARY + "delayed combat" + CC.WHITE
                                + " seen on servers from 2015-2017.",
                        " ",
                        CC.WHITE + "Currently:", this.oldCombat ? CC.GREEN + "Enabled" : CC.RED + "Disabled", " ",
                        CC.BOARD_SEPARATOR, CC.ACCENT + "Click to toggle delayed combat.").build(),
                interaction -> {
                    this.oldCombat = !this.oldCombat;
                    reload();
                });

        setSlot(29, ItemStacks.BACK, interaction -> interaction.getProfile().openMenu(prevMenu));

        setSlot(31, ItemStacks.APPLY, interaction -> {
            val p = interaction.getProfile();
            p.getDuelSettings().setOldCombat(oldCombat);
            p.getDuelSettings().setKnockback(kb == null ? customKnockback : kb);
            p.openMenu(menu);
        });

        setSlot(33,
                ItemStacks.CHOOSE_EXISTING_KNOCKBACK.lore(
                        CC.WHITE + "Choose a " + CC.SECONDARY + "preconfigured knockback"
                                + CC.WHITE + ".",
                        " ",
                        CC.WHITE + "Currently: " + CC.GOLD + (kb == null ? "Custom"
                                : kb.getName()),
                        " ",
                        CC.BOARD_SEPARATOR, CC.ACCENT + "Click to select knockback.").build(),
                interaction -> {

                    int lastKbIndex = this.kbIndex;

                    for (int i = 0; i <= profiles.length; i++) {
                        kbIndex++;

                        if (kbIndex >= profiles.length || kbIndex < 0)
                            kbIndex = 0;

                        if (profiles[kbIndex].getName().contains("combo_kb"))
                            continue;
                        break;
                    }

                    if (lastKbIndex != kbIndex && kbIndex >= 0 && kbIndex < profiles.length) {
                        double friction = (double) profiles[kbIndex].getConfigValues().getOrDefault("friction", customKnockback.getFriction());
                        double horizontal = (double) profiles[kbIndex].getConfigValues().getOrDefault("horizontal", customKnockback.getHorizontal());
                        double vertical = (double) profiles[kbIndex].getConfigValues().getOrDefault("vertical", customKnockback.getVertical());
                        double horizontalExtra = (double) profiles[kbIndex].getConfigValues().getOrDefault("horizontalExtra", customKnockback.getHorizontalExtra());
                        double verticalExtra = (double) profiles[kbIndex].getConfigValues().getOrDefault("verticalExtra", customKnockback.getVerticalExtra());
                        double verticalLimit = (double) profiles[kbIndex].getConfigValues().getOrDefault("verticalLimit", customKnockback.getVerticalLimit());
                        boolean oldCombat = (boolean) profiles[kbIndex].getConfigValues().getOrDefault("oldCombat", duelSettings.isOldCombat());
                        customKnockback.setFriction(friction);
                        customKnockback.setHorizontal(horizontal);
                        customKnockback.setVertical(vertical);
                        customKnockback.setHorizontalExtra(horizontalExtra);
                        customKnockback.setVerticalExtra(verticalExtra);
                        customKnockback.setVerticalLimit(verticalLimit);
                        this.oldCombat = oldCombat;
                    }

                    reload();
                });
    }

    @Override
    public String getTitle() {
        return CC.BLUE + "Create Custom Knockback";
    }

    @Override
    public boolean shouldUpdate() {
        return true;
    }

}

File: src/main/kotlin/gg/mineral/practice/inventory/menus/CreateCustomKnockbackMenu.kt
package gg.mineral.practice.inventory.menus

import gg.mineral.practice.inventory.ClickCancelled
import gg.mineral.practice.inventory.Interaction
import gg.mineral.practice.inventory.Menu
import gg.mineral.practice.inventory.PracticeMenu
import gg.mineral.practice.match.knockback.CustomKnockback
import gg.mineral.practice.util.items.ItemStacks
import gg.mineral.practice.util.messages.CC
import gg.mineral.server.combat.KnockbackProfile
import gg.mineral.server.combat.KnockbackProfileList
import java.math.RoundingMode
import java.text.DecimalFormat

@ClickCancelled(true)
class CreateCustomKnockbackMenu(private val menu: Menu, private val prevMenu: Menu) :
    PracticeMenu() {
    private val profiles: Array<KnockbackProfile> = KnockbackProfileList.getProfiles().values
        .sortedWith { profile1, profile2 ->
            if ("default_kb" == profile1.name) return@sortedWith -1
            if ("default_kb" == profile2.name) return@sortedWith 1
            0
        }.toTypedArray()
    private var kbIndex = 0
    private var oldCombat = false
    private var customKnockback: CustomKnockback? = null

    private fun setFriction(friction: Double) {
        kbIndex = -1
        customKnockback?.friction = friction
    }

    private fun setHorizontal(horizontal: Double) {
        kbIndex = -1
        customKnockback?.horizontal = horizontal
    }

    private fun setVertical(vertical: Double) {
        kbIndex = -1
        customKnockback?.vertical = vertical
    }

    private fun setHorizontalExtra(horizontalExtra: Double) {
        kbIndex = -1
        customKnockback?.horizontalExtra = horizontalExtra
    }

    private fun setVerticalExtra(verticalExtra: Double) {
        kbIndex = -1
        customKnockback?.verticalExtra = verticalExtra
    }

    private fun setVerticalLimit(verticalLimit: Double) {
        kbIndex = -1
        customKnockback?.verticalLimit = verticalLimit
    }

    private fun getFriction(kb: KnockbackProfile): Double {
        if (kbIndex > 0) return kb.configValues.getOrDefault("friction", customKnockback!!.friction) as Double
        return customKnockback!!.friction
    }

    private fun getHorizontal(kb: KnockbackProfile): Double {
        if (kbIndex > 0) return kb.configValues.getOrDefault("horizontal", customKnockback!!.horizontal) as Double
        return customKnockback!!.horizontal
    }

    private fun getVertical(kb: KnockbackProfile): Double {
        if (kbIndex > 0) return kb.configValues.getOrDefault("vertical", customKnockback!!.vertical) as Double
        return customKnockback!!.vertical
    }

    private fun getHorizontalExtra(kb: KnockbackProfile): Double {
        if (kbIndex > 0) return kb.configValues.getOrDefault(
            "horizontalExtra",
            customKnockback!!.horizontalExtra
        ) as Double
        return customKnockback!!.horizontalExtra
    }

    private fun getVerticalExtra(kb: KnockbackProfile): Double {
        if (kbIndex > 0) return kb.configValues.getOrDefault("verticalExtra", customKnockback!!.verticalExtra) as Double
        return customKnockback!!.verticalExtra
    }

    private fun getVerticalLimit(kb: KnockbackProfile): Double {
        if (kbIndex > 0) return kb.configValues.getOrDefault("verticalLimit", customKnockback!!.verticalLimit) as Double
        return customKnockback!!.verticalLimit
    }

    override fun update() {
        val duelSettings = viewer.duelSettings

        for (i in 0..profiles.size) {
            if (kbIndex >= 0 && profiles[kbIndex].name.contains("combo_kb")) {
                kbIndex++
                if (kbIndex >= profiles.size || kbIndex < 0) kbIndex = 0
                continue
            }
            break
        }

        val kb = if (kbIndex >= 0 && kbIndex < profiles.size) profiles[kbIndex] else null

        if (customKnockback == null) {
            customKnockback = if (duelSettings.knockback is CustomKnockback)
                CustomKnockback(duelSettings.knockback as CustomKnockback)
            else
                CustomKnockback()

            if (kbIndex >= 0 && kbIndex < profiles.size) {
                val friction = profiles[kbIndex].configValues.getOrDefault("friction", customKnockback!!.friction)
                val horizontal = profiles[kbIndex].configValues.getOrDefault("horizontal", customKnockback!!.horizontal)
                val vertical = profiles[kbIndex].configValues.getOrDefault("vertical", customKnockback!!.vertical)
                val horizontalExtra =
                    profiles[kbIndex].configValues.getOrDefault("horizontalExtra", customKnockback!!.horizontalExtra)
                val verticalExtra =
                    profiles[kbIndex].configValues.getOrDefault("verticalExtra", customKnockback!!.verticalExtra)
                val verticalLimit =
                    profiles[kbIndex].configValues.getOrDefault("verticalLimit", customKnockback!!.verticalLimit)
                val oldCombat = profiles[kbIndex].configValues.getOrDefault("oldCombat", duelSettings.oldCombat)
                customKnockback!!.friction = friction as Double
                customKnockback!!.horizontal = horizontal as Double
                customKnockback!!.vertical = vertical as Double
                customKnockback!!.horizontalExtra = horizontalExtra as Double
                customKnockback!!.verticalExtra = verticalExtra as Double
                customKnockback!!.verticalLimit = verticalLimit as Double
                this.oldCombat = oldCombat as Boolean
            }
        }

        setSlot(
            10, ItemStacks.FRICTION.name(CC.SECONDARY + CC.B + "Friction")
                .lore(
                    CC.WHITE + "The amount " + CC.SECONDARY + "movement speed and direction",
                    CC.WHITE + "influences knockback magnitude.", " ",
                    CC.WHITE + "Currently:",
                    CC.GOLD + DECIMAL_FORMAT.format(getFriction(customKnockback!!)),
                    CC.BOARD_SEPARATOR, CC.ACCENT + "Click to change value."
                )
                .build()
        ) { interaction: Interaction ->
            interaction.profile
                .openMenu(
                    ConfigureValueMenu.of(
                        this,
                        { friction: Double -> this.setFriction(friction) },
                        Double::class.javaPrimitiveType!!
                    )
                )
        }

        setSlot(
            11, ItemStacks.HORIZONTAL.name(CC.SECONDARY + CC.B + "Horizontal")
                .lore(
                    CC.WHITE + "The " + CC.SECONDARY + "horizontal" + CC.WHITE + " knockback.", " ",
                    CC.WHITE + "Currently:",
                    CC.GOLD + DECIMAL_FORMAT.format(getHorizontal(customKnockback!!)),
                    CC.BOARD_SEPARATOR, CC.ACCENT + "Click to change value."
                )
                .build()
        ) { interaction: Interaction ->
            interaction.profile
                .openMenu(
                    ConfigureValueMenu.of(
                        this,
                        { horizontal: Double -> this.setHorizontal(horizontal) },
                        Double::class.javaPrimitiveType!!
                    )
                )
        }

        setSlot(
            12, ItemStacks.VERTICAL.name(CC.SECONDARY + CC.B + "Vertical")
                .lore(
                    CC.WHITE + "The " + CC.SECONDARY + "vertical" + CC.WHITE + " knockback.", " ",
                    CC.WHITE + "Currently:",
                    CC.GOLD + DECIMAL_FORMAT.format(getVertical(customKnockback!!)),
                    CC.BOARD_SEPARATOR, CC.ACCENT + "Click to change value."
                )
                .build()
        ) { interaction: Interaction ->
            interaction.profile
                .openMenu(
                    ConfigureValueMenu.of(
                        this,
                        { vertical: Double -> this.setVertical(vertical) },
                        Double::class.javaPrimitiveType!!
                    )
                )
        }

        setSlot(
            13, ItemStacks.EXTRA_HORIZONTAL
                .name(CC.SECONDARY + CC.B + "Extra Horizontal")
                .lore(
                    CC.WHITE + "The " + CC.SECONDARY + "horizontal knockback" + CC.WHITE + " added",
                    CC.WHITE + "when sprinting/sprint resetting.", " ",
                    CC.WHITE + "Currently:",
                    CC.GOLD + DECIMAL_FORMAT.format(getHorizontalExtra(customKnockback!!)),
                    CC.BOARD_SEPARATOR, CC.ACCENT + "Click to change value."
                )
                .build()
        ) { interaction: Interaction ->
            interaction.profile
                .openMenu(
                    ConfigureValueMenu.of(
                        this,
                        { horizontalExtra: Double -> this.setHorizontalExtra(horizontalExtra) },
                        Double::class.javaPrimitiveType!!
                    )
                )
        }

        setSlot(
            14, ItemStacks.EXTRA_VERTICAL.name(CC.SECONDARY + CC.B + "Extra Vertical")
                .lore(
                    CC.WHITE + "The " + CC.SECONDARY + "vertical knockback" + CC.WHITE + " added",
                    CC.WHITE + "when sprinting/sprint resetting.", " ",
                    CC.WHITE + "Currently:",
                    CC.GOLD + DECIMAL_FORMAT.format(getVerticalExtra(customKnockback!!)),
                    CC.BOARD_SEPARATOR, CC.ACCENT + "Click to change value."
                )
                .build()
        ) { interaction: Interaction ->
            interaction.profile
                .openMenu(
                    ConfigureValueMenu.of(
                        this,
                        { verticalExtra: Double -> this.setVerticalExtra(verticalExtra) },
                        Double::class.javaPrimitiveType!!
                    )
                )
        }

        setSlot(
            15, ItemStacks.VERTICAL_LIMIT.name(CC.SECONDARY + CC.B + "Vertical Limit")
                .lore(
                    CC.WHITE + "The " + CC.SECONDARY + "limit" + CC.WHITE + " to vertical knockback.",
                    " ",
                    CC.WHITE + "Currently:",
                    CC.GOLD + DECIMAL_FORMAT.format(getVerticalLimit(customKnockback!!)),
                    CC.BOARD_SEPARATOR, CC.ACCENT + "Click to change value."
                )
                .build()
        ) { interaction: Interaction ->
            interaction.profile
                .openMenu(
                    ConfigureValueMenu.of(
                        this,
                        { verticalLimit: Double -> this.setVerticalLimit(verticalLimit) },
                        Double::class.javaPrimitiveType!!
                    )
                )
        }

        setSlot(
            16, ItemStacks.OLD_COMBAT.name(CC.SECONDARY + CC.B + "Delayed Combat").lore(
                (CC.WHITE + "Play using " + CC.SECONDARY + "delayed combat" + CC.WHITE
                        + " seen on servers from 2015-2017."),
                " ",
                CC.WHITE + "Currently:", if (this.oldCombat) CC.GREEN + "Enabled" else CC.RED + "Disabled", " ",
                CC.BOARD_SEPARATOR, CC.ACCENT + "Click to toggle delayed combat."
            ).build()
        ) {
            this.oldCombat =
                !this.oldCombat
            reload()
        }

        setSlot(
            29, ItemStacks.BACK
        ) { interaction: Interaction -> interaction.profile.openMenu(prevMenu) }

        setSlot(31, ItemStacks.APPLY) { interaction: Interaction ->
            val p = interaction.profile
            p.duelSettings.oldCombat = oldCombat
            p.duelSettings.knockback = kb ?: customKnockback
            p.openMenu(menu)
        }

        setSlot(
            33,
            ItemStacks.CHOOSE_EXISTING_KNOCKBACK.lore(
                (CC.WHITE + "Choose a " + CC.SECONDARY + "preconfigured knockback"
                        + CC.WHITE + "."),
                " ",
                CC.WHITE + "Currently: " + CC.GOLD + (if (kb == null)
                    "Custom"
                else
                    kb.name),
                " ",
                CC.BOARD_SEPARATOR, CC.ACCENT + "Click to select knockback."
            ).build()
        ) {
            val lastKbIndex = this.kbIndex
            for (i in 0..profiles.size) {
                kbIndex++

                if (kbIndex >= profiles.size || kbIndex < 0) kbIndex = 0

                if (profiles[kbIndex].name.contains("combo_kb")) continue
                break
            }

            if (lastKbIndex != kbIndex && kbIndex >= 0 && kbIndex < profiles.size) {
                val friction =
                    profiles[kbIndex].configValues.getOrDefault("friction", customKnockback!!.friction) as Double
                val horizontal = profiles[kbIndex].configValues.getOrDefault(
                    "horizontal",
                    customKnockback!!.horizontal
                ) as Double
                val vertical =
                    profiles[kbIndex].configValues.getOrDefault("vertical", customKnockback!!.vertical) as Double
                val horizontalExtra = profiles[kbIndex].configValues.getOrDefault(
                    "horizontalExtra",
                    customKnockback!!.horizontalExtra
                ) as Double
                val verticalExtra = profiles[kbIndex].configValues.getOrDefault(
                    "verticalExtra",
                    customKnockback!!.verticalExtra
                ) as Double
                val verticalLimit = profiles[kbIndex].configValues.getOrDefault(
                    "verticalLimit",
                    customKnockback!!.verticalLimit
                ) as Double
                val oldCombat =
                    profiles[kbIndex].configValues.getOrDefault("oldCombat", duelSettings.oldCombat) as Boolean
                customKnockback!!.friction = friction
                customKnockback!!.horizontal = horizontal
                customKnockback!!.vertical = vertical
                customKnockback!!.horizontalExtra = horizontalExtra
                customKnockback!!.verticalExtra = verticalExtra
                customKnockback!!.verticalLimit = verticalLimit
                this.oldCombat = oldCombat
            }
            reload()
        }
    }

    override val title: String
        get() = CC.BLUE + "Create Custom Knockback"

    override fun shouldUpdate() = true

    companion object {
        private val DECIMAL_FORMAT = DecimalFormat("#.###")

        init {
            DECIMAL_FORMAT.roundingMode = RoundingMode.HALF_DOWN
        }
    }
}
